docs/proposals/post-deployment-hooks.md

```diff
@@ -0,0 +1,223 @@
+
+Deployment hooks are implemented as run-once pods which can be executed at one or both of the following points during the deployment lifecycle:
+
+1. Before the execution of the deployment strategy, which ensures that the hook is run and its outcome evaluated prior to the scale-up of the deployment. These hooks are referred to as *pre-deployment* hooks.
+2. After the execution of the deployment strategy. These hooks are referred to as *post-deployment* hooks.
+
+##### Hook failure handling
+
+Hooks designated as *mandatory* should impact the outcome of the deployment.
+
+There are a few possible ways to handle a failed mandatory pre-deployment hook:
+
+1. Transition the deployment to a failed status and do not execute the strategy.
+2. Delete and retry the deployment.
+  1. Potentially safe because the strategy has not yet executed and the existing prior deployment is still active.
+    1. Could be unsafe depending on what the hook did prior to failing.
+  2. Further API considerations need made to prevent endless deployment attempts.
+3. Rollback the deployment to a previous version.
+  1. This is very dangerous to do automatically and is probably not realistic at this time:
+    1. Requires automated rollback viability analysis/trust.
+    2. Requires logic to prevent a chronically failing hook that spans historical deployments from causing unending rollbacks.
+
+This proposal prescribes the use of option 1 as being the simplest starting point for the hook API.
+
+Failed mandatory post-deployment hooks are more challenging:
+
+1. The deployment has most likely already been rolled out and made live by the strategy.
+2. Deleting the deployment is no longer safe due to 1.
+3. Rollback is necessary and subject to the same challenges presented above.
+
+Due to the complexities of automated rollback, this proposal limits the scope of failure handling for post-deployment hooks: post-deployment hooks cannot be considered mandatory at this time. This limitation may be lifted in the future by an separate proposal.
+
+##### Hook failure reporting
+
+When a deployment hook fails:
+
+1. An error is logged via the global error handler.
+2. An event will be fired.
+
+More reporting capabilities could be addressed in a future proposal.
+
+
+### Deployment hook API resource
+
+A new API resource will represent a hook:
+
+```go
+// DeploymentHook describes a run-once pod which can execute
+// a deployment hook. If the hook is Mandatory, the outcome
+// of the hook can affect the deployment lifecycle.
+//
+// The template's RestartPolicy may be either RestartPolicyNever
+// or RestartPolicyOnFailure. The default is RestartPolicyNever.
+//
+// The RestartPolicy limitation:
+//  * Enforces run-once semantics
+//  * Provides the hook user with retry capabilities
+type DeploymentHook struct {
+  // Template describes a run-once pod which executes and
+  // represents a hook.
+  Template *PodTemplate
+
+  // Mandatory indicates whether a failed hook execution
```

Comment:
Also how would I specify to use the same image I jsit deployed?

To me the run template should be the exact pod and maybe allow overriding the command (like a lifecycle hook).  Or you run the pod with only one container.  I don't see how the pod template makes this easy for the simple case